### PR TITLE
fix: Add SIGTERM handling

### LINKED ISSUE #295: [feat] Add SIGTERM handling to `scripts/data/client.py`
Currently it takes several Ctrl-C to stop scripts/data/client.py:
image
Add proper SIGTERM handling to the script so it behaves as expected. Make sure no unnecessary exceptions are written to the logs.

### PR DESCRIPTION
**Pull request by raizo07**
- [x] resolves #295 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

Add SIGTERM handling to scripts/data/client.py

### CODE REVIEW

**Comment by raizo07** [comment]
@maciejka It exits correctly.
image

File: scripts/data/client.py [hunk 1]
```diff
@@ -152,52 +155,49 @@ def job_producer(job_gen):
                     and current_weight != 0
                     or job_queue.full()
                 ):
+                    if shutdown_event.is_set():
+                        logger.info("Shutdown event detected while waiting. Exiting...")
+                        return
+
                     logger.debug("Producer is waiting for weight to be released.")
-                    weight_lock.wait()  # Wait for the condition to be met
+                    weight_lock.wait()
```
Comment (maciejka) [inline scripts/data/client.py:163]: What will happen when SIGTERM is sent when thread is waiting for a lock?

Comment (raizo07) [reply scripts/data/client.py:163]: Threads waiting on the lock will wait until they are notified (either by a timeout or by other threads) before they can check the `shutdown_event` and exit.

Comment (maciejka) [reply scripts/data/client.py:163]: This means SIGTERM will not terminate the program immediately?

Comment (raizo07) [reply scripts/data/client.py:163]: Yes, it will not terminate the program immediately rather it initiates a graceful shutdown process which will allow active threads to finish their current tasks and lets waiting threads proceed once they acquire the lock.

Comment (maciejka) [reply scripts/data/client.py:163]: The point of sending SIGTERM is to stop immediately.

File: scripts/data/client.py [hunk 2]
```diff
@@ -16,32 +15,51 @@
...
 QUEUE_MAX_SIZE = THREAD_POOL_SIZE * 2  # Maximum size of the job queue
-
 BASE_DIR = Path(".client_cache")
 
 # Shared state variables
 current_weight = 0
 weight_lock = threading.Condition()
 job_queue = queue.Queue(maxsize=QUEUE_MAX_SIZE)
 
+shutdown_event = threading.Event()
```
Comment (maciejka) [inline scripts/data/client.py:34]: does this need to be an event? No one waits for it.

File: scripts/data/client.py [hunk 3]
```diff
@@ -16,32 +15,51 @@
...
+executor = None  # Global executor reference for shutdown
+
+
+def handle_sigterm(signum, frame):
+    """Handle SIGTERM by initiating immediate shutdown"""
+    logger.info("Received SIGTERM signal. Initiating immediate shutdown...")
+    shutdown_event.set()
+
+    # Clear the job queue
+    while not job_queue.empty():
```
Comment (maciejka) [inline scripts/data/client.py:44]: signal handler modifying `job_queue` does not look like a good design. I would just set `shutdown_event` here. 
In `job_producer` when shutdown is detected all consumers can be notified with `notify_all`. Cleaning `job_queue` does not seem necessary.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: scripts/data/client.py (hunk 2 above)
Comment (raizo07) [reply scripts/data/client.py:33]: Yes it does, because it is checked in multiple threads and also a part of the shutdown signal system

File: scripts/data/client.py [hunk 4]
```diff
@@ -125,101 +149,104 @@ def process_batch(job):
...
-
-                # Notify consumers that a new job is available
                 weight_lock.notify_all()
+
     finally:
         logger.debug("Producer is exiting...")
-        for _ in range(THREAD_POOL_SIZE):
-            job_queue.put(None)
-
+        if not shutdown_event.is_set():
```
Comment (maciejka) [inline scripts/data/client.py:208]: Why is it necessary? If there are no more jobs we just let consumers finish, no?

File: scripts/data/client.py [hunk 5]
```diff
@@ -125,101 +149,104 @@ def process_batch(job):
...
 
 
-# Producer function: Generates data and adds jobs to the queue
 def job_producer(job_gen):
     global current_weight
 
     try:
         for job, weight in job_gen:
-            # Wait until there is enough weight capacity to add the new block
+            if shutdown_event.is_set():
```
Comment (maciejka) [inline scripts/data/client.py:181]: Is this check necessary? `shutdown_event` is checked few lines below.

File: scripts/data/client.py [hunk 6]
```diff
@@ -56,41 +64,44 @@ def __str__(self):
...
     end = start + blocks
-
     height_range, step = (
         ([random.randint(start, end) for _ in range(start, end)], 1)
         if strategy == "random"
         else (range(start, end, step), step)
     )
-
     for height in height_range:
+        if shutdown_event.is_set():
```
Comment (maciejka) [inline scripts/data/client.py:87]: No need to check `shutdown_event` here, it is checked in `job_producer`:
image

File: scripts/data/client.py (hunk 4 above)
Comment (maciejka) [reply scripts/data/client.py:208]: If you let this happen during shutdown you will not need to wait with timeout in line 224. Information that there is no more work will reach consumers.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: scripts/data/client.py [hunk 7]
```diff
@@ -12,36 +11,45 @@
...
+# Initialize logger
 logger = logging.getLogger(__name__)
 
 # Constants
 MAX_WEIGHT_LIMIT = 8000  # Total weight limit for all jobs
 THREAD_POOL_SIZE = os.cpu_count()  # Number of threads for processing
 QUEUE_MAX_SIZE = THREAD_POOL_SIZE * 2  # Maximum size of the job queue
-
-BASE_DIR = Path(".client_cache")
+BASE_DIR = Path(__file__).resolve().parent / ".client_cache"  # Use absolute path
```
Comment (maciejka) [inline scripts/data/client.py:27]: Why this change is necessary?

File: scripts/data/client.py [hunk 8]
```diff
@@ -56,41 +64,41 @@ def __str__(self):
...
 
-# Function to process a batch
-def process_batch(job):
-    arguments_file = job.batch_file.as_posix().replace(".json", "-arguments.json")
-
-    with open(arguments_file, "w") as af:
-        af.write(str(format_args(job.batch_file, False, False)))
-
-    result = subprocess.run(
+def run_process(arguments_file):
```
Comment (maciejka) [inline scripts/data/client.py:99]: It should take whole command as an argument

File: scripts/data/client.py [hunk 9]
```diff
@@ -102,124 +110,132 @@ def process_batch(job):
...
-        gas_info = f"gas spent: {int(match.group(1))}" if match else "no gas info found"
-        logger.info(f"{job} done, {gas_info}")
-        if not match:
-            logger.warning(f"{job}: not gas info found")
+    except Exception as e:
+        logger.error(f"Error processing batch {job}: {e}")
 
 
-# Producer function: Generates data and adds jobs to the queue
 def job_producer(job_gen):
```
Comment (maciejka) [inline scripts/data/client.py:180]: Why do you remove comments?

File: scripts/data/client.py [hunk 10]
```diff
@@ -102,124 +110,132 @@ def process_batch(job):
...
     while True:
         try:
-            logger.debug(
-                f"Consumer is waiting for a job. Queue length: {job_queue.qsize()}"
-            )
-            # Get a job from the queue
-            work_to_do = job_queue.get(block=True)
+            work_to_do = job_queue.get()
 
             if work_to_do is None:
```
Comment (maciejka) [inline scripts/data/client.py:219]: Why do you remove log here?

File: scripts/data/client.py [hunk 11]
```diff
@@ -102,124 +110,132 @@ def process_batch(job):
...
-                logger.debug("No more work to do, consumer is exiting.")
                 job_queue.task_done()
                 break
 
-            (job, weight) = work_to_do
-
-            # Process the block
+            job, weight = work_to_do
             try:
```
Comment (maciejka) [inline scripts/data/client.py:224]: Why log is removed here?

File: scripts/data/client.py [hunk 12]
```diff
@@ -102,124 +110,132 @@ def process_batch(job):
...
-            with weight_lock:
-                current_weight -= weight
-                logger.debug(
-                    f"Finished processing job, current total weight: {current_weight}"
-                )
-                weight_lock.notify_all()  # Notify producer to add more jobs
-
-            # Mark job as done
-            job_queue.task_done()
+            finally:
```
Comment (maciejka) [inline scripts/data/client.py:226]: Why `except` block is removed? Exceptions might happen during execution. You probably need to handle process termination exception explicitly.

**Review by maciejka — CHANGES_REQUESTED** [review]

**Comment by maciejka** [comment]
Hey @raizo07 there are many unanswered comments. Please respond to all of them.

**Comment by raizo07** [comment]
@maciejka Okay, I'll respond to them

File: scripts/data/client.py [hunk 13]
```diff
@@ -56,41 +64,44 @@ def __str__(self):
...
 
-# Function to process a batch
-def process_batch(job):
-    arguments_file = job.batch_file.as_posix().replace(".json", "-arguments.json")
-
-    with open(arguments_file, "w") as af:
-        af.write(str(format_args(job.batch_file, False, False)))
-
-    result = subprocess.run(
+def run_process(arguments_file, job):
```
[the remark this answers is not in the document]
Comment (raizo07) [reply scripts/data/client.py]: I've refactored `run_process` and `process_batch` to better separate concerns.

File: scripts/data/client.py (hunk 5 above)
Comment (raizo07) [reply scripts/data/client.py:213]: I've removed redundant `shutdown` checks in `job_generator`

File: scripts/data/client.py (hunk 4 above)
Comment (raizo07) [reply scripts/data/client.py]: I've refactored Producer to now properly signal end of work to all consumers using None sentinels

File: scripts/data/client.py (hunk 7 above)
Comment (raizo07) [reply scripts/data/client.py]: Although I've reverted the change, but using an absolute path for `BASE_DIR` is better than a relative path because it ensures consistent behaviour regardless of where the script is executed from.

File: scripts/data/client.py (hunk 8 above)
Comment (raizo07) [reply scripts/data/client.py]: Resolved in the latest fix.

File: scripts/data/client.py (hunk 9 above)
Comment (raizo07) [reply scripts/data/client.py]: An oversight from my end, reverted to the previous form with the comment in the latest fix.

File: scripts/data/client.py (hunk 10 above)
Comment (raizo07) [reply scripts/data/client.py:273]: Resolved.

File: scripts/data/client.py (hunk 11 above)
Comment (raizo07) [reply scripts/data/client.py]: Resolved and properly handled in the latest pr

File: scripts/data/client.py (hunk 12 above)
Comment (raizo07) [reply scripts/data/client.py]: I've handled process termination exception explicitly as suggested in the latest pr

**Comment by raizo07** [comment]
@maciejka kindly review for any further changes

File: scripts/data/client.py [hunk 14]
```diff
@@ -56,170 +74,193 @@ def __str__(self):
...
             batch_weight = calculate_batch_weight(batch_data, mode)
             yield Job(height, step, mode, batch_weight, batch_file), batch_weight
         except Exception as e:
             logger.error(f"Error while generating data for: {height}:\n{e}")
 
 
-# Function to process a batch
 def process_batch(job):
-    arguments_file = job.batch_file.as_posix().replace(".json", "-arguments.json")
+    if shutdown_event.is_set():
```
[the remark this answers is not in the document]
Comment (maciejka) [reply scripts/data/client.py:134]: Why this is resolved? `job_generator` and `process_batch` is about bussines logic those functions should not bother with shutdown subtleties.

File: scripts/data/client.py (hunk 6 above)
Comment (maciejka) [reply scripts/data/client.py:114]: Why this is resolved? job_generator and process_batch is about bussines logic those functions should not bother with shutdown subtleties.

**Review by maciejka — CHANGES_REQUESTED** [review]

**Comment by raizo07** [comment]
@maciejka The early check allows the producer to quickly exit if a shutdown is requested before processing a job.

**Comment by raizo07** [comment]
Hey @maciejka I saw you replaced CancellationToken which I introduced with shared variables? 

For the sake of learning, why is that?

**Comment by maciejka** [comment]
Just simpler
